File: 15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */

// @lc code=start
class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;

        if (nums.size() < 3)
            return {};

        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size() - 2; i++) {
            while (i > 0 && i < nums.size() - 2 && nums[i - 1] == nums[i]) {
                i++;
            }

            int target = -nums[i];

            int l = i + 1;
            int r = nums.size() - 1;

            while (l < r) {
                int sum = nums[l] + nums[r];

                if (sum < target)
                    while (nums[++l] == nums[l - 1] && l < r) {
                    }

                else if (sum > target && r > l)
                    while (nums[--r] == nums[r + 1] && r > l) {
                    }

                else {
                    result.push_back(vector<int>{nums[i], nums[l], nums[r]});
                    while (nums[++l] == nums[l - 1] && l < r) {
                    }
                    while (nums[--r] == nums[r + 1] && r > l) {
                    }
                }
            }
        }

        return result;
    }
};

// @lc code=end

vector<vector<int>> threeSum(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    int n = nums.size();
    vector<vector<int>> result;
    for (int i = 0; i < nums.size() - 2; i++) {
        if (i > 0 && nums[i - 1] == nums[i]) {
            continue;
        }

        int left = i + 1, right = n - 1;

        while (left < right) {
            int sum = nums[i] + nums[left] + nums[right];

            if (sum < 0)
                left++;
            else if (sum > 0)
                right--;
            else {
                result.push_back({nums[i], nums[left], nums[right]});
                while (left + 1 < right && nums[left] == nums[left + 1]) left++;
                while (left < right - 1 && nums[right] == nums[right - 1]) right--;
                left++;
                right++;
            }
        }
    }

    return result;
}

class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        for (int i = 0; i < nums.size(); i++) {
            if (i > 0 && nums[i - 1] == nums[i]) {
                continue;
            }
            if (nums[i] > 0) break;
            for (int j = i + 1; j < nums.size(); j++) {
                for (int k = nums.size() - 1; k > j; k--) {
                    if (nums[i] + nums[j] + nums[k] == 0) {
                        vector<int> temp = {nums[i], nums[j], nums[k]};
                        if (result.size() == 0 || temp != result[result.size() - 1])
                            result.push_back(temp);
                        break;
                    }

                    if (nums[i] + nums[j] + nums[k] < 0)
                        break;
                }
            }
        }

        return result;
    }
};

class Solution {
   public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if (nums.size() < 3) return {};

        vector<vector<int>> result;

        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                for (int k = j + 1; k < nums.size(); k++) {
                    if (nums[i] + nums[j] + nums[k] == 0) {
                        int a = max(nums[i], max(nums[j], nums[k]));
                        int c = min(nums[i], min(nums[j], nums[k]));
                        int b = 0 - a - c;

                        vector<int> temp = {a, b, c};
                        // printf("{%i, %i, %i}\n", a, b, c);

                        bool exists = false;
                        for (auto v : result) {
                            if (v == temp)
                                exists = true;
                        }
                        if (!exists)
                            result.push_back(temp);
                    }
                }
            }
        }

        return result;
    }
};

File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
class Solution:
    """
    Best conceivable runtime: O(n^2)
    """
    """
    Brute force: generate all possible triplets
    """
    # skip

    """
    Hashset: O(n^2)
    
    """
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         result = []
#         nums.sort()
#         for i in range(len(nums)):
#             if nums[i] > 0:
#                 break
#             if i == 0 or nums[i] != nums[i-1]:
#                 self.find_triple(nums, i, result)
#         return result

#     def find_triple(self, nums, i, result): # find triple starting with element i
#         seen = set()
#         j = i + 1
#         while j < len(nums):
#             complement = 0 - nums[i] - nums[j]
#             if complement in seen:  # found triplet!
#                 result.append([nums[i], complement, nums[j]])
#                 while j + 1 < len(nums) and nums[j] == nums[j+1]:
#                     j += 1
#             seen.add(nums[j])
#             j += 1

    """
    Two pointers -> O(n^2)
    """
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         result = []
#         nums.sort()
#         for i in range(len(nums)):
#             if nums[i] > 0:
#                 break
#             if i == 0 or nums[i] != nums[i-1]:
#                 self.find_triple(nums, i, result)
#         return result

#     def find_triple(self, nums, i, result): # find triple starting with element i
#         target_sum = 0 - nums[i]
#         left, right = i + 1, len(nums) - 1
#         while left < right:
#             curr_sum = nums[left] + nums[right]
#             if curr_sum == target_sum:
#                 result.append([nums[i], nums[left], nums[right]])
#                 left += 1
#                 right -= 1
#                 while left < right and nums[left] == nums[left - 1]:
#                     left += 1
#             elif curr_sum < target_sum:
#                 left += 1
#             else:
#                 right -= 1

    """
    인터뷰에서는 중요하지 않음!!!!!
    No-sort: if can't modify input array / can't copy to sort because large size?
    Use HashSet, adapted for unsorted array (put triples in hashset, values in triple ordered)
    """

    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = set()
        duplicates = set()
        seen = {}

        for i, val1 in enumerate(nums):
            if val1 not in duplicates:  # not seen yet
                duplicates.add(val1)
                for j, val2 in enumerate(nums[i+1:]):
                    complement = -val1 - val2
                    # i + 1 ~ j - 1 사이인 원소
                    if complement in seen and seen[complement] == i:
                        result.add(tuple(sorted((val1, val2, complement))))
                    seen[val2] = i  # val2를 마지막으로 확인한 건 i다
        return result

    """
    Failed attempt..
    Sort
    [-1,0,1,2,-1,-4]
    [-4, -1, -1, 0, 1, 2]
    Two pointer + binary search
    For two pointer:
        update pointer (until element with different value) for the element with greater absolute value
    For binary search:
        search for 0 - a - b in the range [left + 1], [right + 1]
    """
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         def binary_search(left: int, right: int, target: int) -> bool:
#             while left <= right:
#                 mid = (left + right) // 2
#                 if nums[mid] == target:
#                     return True
#                 elif target < nums[mid]:
#                     right = mid - 1
#                 else:
#                     left = mid + 1

#             return False


#         left = 0
#         right = len(nums) - 1
#         result = []
#         nums.sort()
#         prev_left = -1
#         while left + 1 < right:
#             # check current pair
#             mid_element = 0 - nums[left] - nums[right]
#             if binary_search(left + 1, right - 1, mid_element):
#                 result.append([nums[left], mid_element, nums[right]])
#             if prev_left >= 0:
#                 left = prev_left
#                 prev_left = -1
#                 while right > left and nums[right] == nums[right - 1]:
#                     right -= 1
#                 right -= 1
#                 continue
#             if -nums[left] == nums[right]:
#                 prev_left = left
#                 while left < right and nums[left] == nums[left + 1]:
#                     left += 1
#                 left += 1
#             elif -nums[left] > nums[right]:
#                 while left < right and nums[left] == nums[left + 1]:
#                     left += 1
#                 left += 1

#             else:
#                 while right > left and nums[right] == nums[right - 1]:
#                     right -= 1
#                 right -= 1
#         return result


# @lc code=end
